pkg/controller/controller_instance.go

```diff
@@ -1227,6 +1237,77 @@ func (c *controller) pollServiceInstance(serviceClass *v1alpha1.ServiceClass, se
+// with the appropriate error message.
+func (c *controller) resolveReferences(instance *v1alpha1.ServiceInstance) (*v1alpha1.ServiceInstance, error) {
+	if instance.Spec.ServiceClassRef != nil && instance.Spec.ServicePlanRef != nil {
+		return instance, nil
+	}
+
+	if instance.Spec.ServiceClassRef == nil {
+		glog.V(4).Infof("looking up a ServiceClass from externalName: %q", instance.Spec.ExternalServiceClassName)
+		listOpts := apimachineryv1.ListOptions{FieldSelector: "spec.externalName==" + instance.Spec.ExternalServiceClassName}
+		serviceClasses, err := c.serviceCatalogClient.ServiceClasses().List(listOpts)
+		if err == nil && len(serviceClasses.Items) == 1 {
+			sc := &serviceClasses.Items[0]
+			instance.Spec.ServiceClassRef = &apiv1.ObjectReference{
+				Kind:            sc.Kind,
+				Namespace:       sc.Namespace,
+				Name:            sc.Name,
+				UID:             sc.UID,
+				APIVersion:      sc.APIVersion,
+				ResourceVersion: sc.ResourceVersion,
+			}
+			glog.V(4).Infof("Resolve ServiceClass %q to %q", instance.Spec.ExternalServiceClassName, instance.Spec.ServiceClassRef)
+		} else {
+			s := fmt.Sprintf("ServiceInstance \"%s/%s\" references a non-existent ServiceClass %q", instance.Namespace, instance.Name, instance.Spec.ExternalServiceClassName)
+			glog.Warning(s)
+			c.updateServiceInstanceCondition(
+				instance,
+				v1alpha1.ServiceInstanceConditionReady,
+				v1alpha1.ConditionFalse,
+				errorNonexistentServiceClassReason,
+				"The instance references a ServiceClass that does not exist. "+s,
+			)
+			c.recorder.Event(instance, api.EventTypeWarning, errorNonexistentServiceClassReason, s)
+			return nil, fmt.Errorf(s)
+		}
+	}
+
+	if instance.Spec.ServicePlanRef == nil {
+		fieldSelector := fields.SelectorFromSet(fields.Set{"spec.externalName": instance.Spec.ExternalServicePlanName, "spec.serviceClassRef.name": instance.Spec.ServiceClassRef.Name}).String()
+		listOpts := apimachineryv1.ListOptions{FieldSelector: fieldSelector}
+		servicePlans, err := c.serviceCatalogClient.ServicePlans().List(listOpts)
+		if err == nil && len(servicePlans.Items) == 1 {
+			sp := &servicePlans.Items[0]
+			instance.Spec.ServicePlanRef = &apiv1.ObjectReference{
+				Kind:            sp.Kind,
+				Namespace:       sp.Namespace,
+				Name:            sp.Name,
+				UID:             sp.UID,
+				APIVersion:      sp.APIVersion,
+				ResourceVersion: sp.ResourceVersion,
+			}
+			glog.V(4).Infof("Resolve ServicePlan %q to %q", instance.Spec.ExternalServicePlanName, instance.Spec.ServicePlanRef)
+		} else {
+			s := fmt.Sprintf("ServiceInstance \"%s/%s\" references a non-existent ServicePlan %q on ServiceClass %q", instance.Namespace, instance.Name, instance.Spec.ExternalServicePlanName, instance.Spec.ExternalServiceClassName)
+			glog.Warning(s)
+			c.updateServiceInstanceCondition(
+				instance,
+				v1alpha1.ServiceInstanceConditionReady,
+				v1alpha1.ConditionFalse,
+				"ReferencesNonexistentServicePlan",
```

Comment:
I think we should set this as a constant, like `errorNonexistentServiceClassReason` above.

Reply:
done.